### ISSUE TITLE
Shared configuration not applied when using InheritInverseConfiguration

### ISSUE THREAD

**Issue by mricciuti** [mapstruct/mapstruct#1367]
I am trying to make both features "Shared Configuration" and "Inverse Mapping" work together, without success. ( Version used:  1.2.0.Final )
I use MapStruct in my project in order to handle mapping between domain entities layer (JPA entities) and REST resources layer (simple DTOs used by REST/JSON API).
- shared configuration:
Both entities and dtos layers have a root base class that handle common properties, so I have implemented a MapperConfig to handle mapping rules for these fields : 
```
@MapperConfig(unmappedTargetPolicy = ReportingPolicy.WARN,
        mappingInheritanceStrategy = MappingInheritanceStrategy.AUTO_INHERIT_ALL_FROM_CONFIG)
public interface EntityToDtoMappingConfig {
    
    @Mapping(target = "dbId", source = "id")
    @Mapping(target = "links", ignore = true)
    BaseDto entityToDto(final BaseEntity entity);
    
    @Mapping(target = "id", source = "dbId")
    @Mapping(target = "createdBy", ignore = true)
    @Mapping(target = "creationDate", ignore = true)
    @Mapping(target = "lastModifiedBy", constant = "restApiUser") // force modifiedBy with restApiUser constant
    @Mapping(target = "lastModifiedDate", ignore = true)
    public BaseEntity dtoToEntity(BaseDto dto);
    
}
```
- Inverse Mapping
I use @InheritInverseConfiguration in order to simplify "mapFrom" methods definition: generally all mapping rules are defined on the "mapTo" methods, so "mapFrom" methods can be defined simply with "inverse mapping" :
```
@Mapper(config = EntityToDtoMappingConfig.class)
public abstract class CarMapper {

    @Mapping(source = "manufacturer", target = "maker")
    @Mapping(source = "numberOfSeats", target = "seatCount")
    abstract public CarDto mapTo(Car car);
    
    @InheritInverseConfiguration(name = "mapTo")
    // inheritInvers configuration should map both maker and seatCount properties.
    // additional mapping should also be inherited from   EntityToDtoMappingConfig.dtoToEntity method, BUT IT DOES NOT SEEM TO WORK
    abstract Car mapFrom(CarDto carDto);
    
}
```

Mapping from Entity to DTO is working fine: inheritance is handled correctly and mapping rules from EntityToDtoMappingConfig  are appended to mapping rules defined on CarMapper.mapTo method.

But mapping from DTO to Entity is not working as expected :  mapping rules defined in the EntityToDtoMappingConfig  are not applied.

Question : is there a limitation when using InheritInverseConfiguration that prevent us using shared configuration mechanisme?  or is this a bug ?

I have shared a concrete and simple project on GitHub to demonstrate this issue : see  https://github.com/mricciuti/sandbox-mapstruct

Thanks by advance.

**Comment by sjaakd** [comment]
@mricciuti .. I'll have a look. There is however a limitation on how many times you can inherit (to avoid cyclic behaviour).

**Comment by sjaakd** [comment]
The problem is that `MappingInheritanceStrategy.AUTO_INHERIT_ALL_FROM_CONFIG` does not work altogether different from explicit referring to `@InheritInverseConfiguration` or `@InheritConfiguration`. Once you remove the `MappingInheritanceStrategy.AUTO_INHERIT_ALL_FROM_CONFIG` in favour of explicit inheritance you get a clear error message: `Method cannot be annotated with both a @InheritConfiguration and @InheritInverseConfiguration.`. The documentation is not altogether clear on this subject.

@agudian : I think we discussed this when implementing this functionality. At hindsight the use case of @mricciuti makes sense. I remember vaguely that I encountered similar limitations myself. WDYT?:
1. should we relax this case and allow both  `@InheritConfiguration` and `@InheritInverseConfiguration.`?
2. do you foresee any problems here?

**Comment by gunnarmorling** [comment]
@sjaakd, I think my idea back than was that one of the two "forward" mapping methods would inherit its configuration from the other one, and the two inverse mapping methods would each inverse-inherit from one of the two forward mapping methods:

    forwardB -> forwardA
    inverseA -> forwardA
    inverseB -> forwardB

I.e. for inverseB there'd be two levels of inheritance. Would that make sense?

**Comment by sjaakd** [comment]
> I.e. for inverseB there'd be two levels of inheritance. Would that make sense?

If I understand you correctly, something like this:
inverseB -> forwardB -> forwardA
In other words: inverseB would get its mappings from forwardB merged with forwardA? I played around with the tests and this seems indeed to be the case.  

----

Whereas its not possible to combine forward and reverse explicit configurations (you get a warning:  `Method cannot be annotated with both a @InheritConfiguration and @InheritInverseConfiguration`) it seems to be possible with the current auto-inheritance options. Here you have the possibilities: `AUTO_INHERIT_FROM_CONFIG`, `AUTO_INHERIT_REVERSE_FROM_CONFIG`, and **`AUTO_INHERIT_ALL_FROM_CONFIG`** (which seems to work)..

----
To be short: we have the possibility to transitive merge and in the auto-inherit case we can both inherit forward/reverse but **not** in the explicit inherit case.  Hmmm. I don't like this.. This transitive merge is obfuscating stuff (possibly allowing cyclic inheritance).  

We should clean this up in 2.0. At least aligning the explicit case with the auto-inherit case.

----
Ps. Perhaps swap it out for an altogether more simple approach. I still like the idea of having template mapping annotations.

Like:
```java
@MappingDefinition
@Target(ElementType.METHOD)
@Retention(RetentionPolicy.CLASS)
@Mapping( target="key", ignore=true )
@Mapping( target="id", ignore =true )
public @interface EntityMappings {
}
```

**Comment by filiphr** [comment]
@mricciuti Sjaak has provided a PR #1397 for this issue. Is it possible for you to try it out on your codebase?

**Comment by mricciuti** [comment]
Hi, 
I have integrated the PR #1397 and tested on my codebase:  I can confirm that this PR solves my issue, and the mapping rules from both inverse  and inherited configurations are applied.
thanks!

**Comment by sjaakd** [comment]
@mricciuti .. thanks for testing. Merged #1397 to master
